staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go

```diff
@@ -160,3 +172,70 @@ func decorateResponseWriter(responseWriter http.ResponseWriter, out io.Writer, i
 	}
 	return delegate
 }
+
+var _ http.ResponseWriter = &auditResponseWriter{}
+var cloner = conversion.NewCloner()
+
+// auditResponseWriter intercepts WriteHeader, sets it in the event. If the sink is set, it will
+// create immediately an event (for long running requests).
+type auditResponseWriter struct {
+	http.ResponseWriter
+	event *auditinternal.Event
+	once  sync.Once
+	sink  chan<- *auditinternal.Event
+}
+
+func (a *auditResponseWriter) processCode(code int) {
+	a.once.Do(func() {
```

Comment:
> That makes sense, although it doubles the QPS to the audit backend.
> 
> Maybe this should be part of the rule configuration? For instance, it's probably not necessary on GET requests.

I would leave it as a choice for sink implemenations.  I doubt a cluster-admin will ever think to configure it in a sensible way.